### DISCUSSION TITLE
Search Topo Faces and Topo Edges and create Labels

### DISCUSSION THREAD

**Discussion by dkowollik** [ansys/pyprimemesh#864] [Q&A] [answered]
I need to search Topo Faces and Topo Edges and compare them with coordinates to be able to assign labels, which are not transfered from Rhino files. Is it possible to do this within the prime model? If so, can you please provide an example?

**Comment by waltersma** [comment, marked as answer]
Hi @dkowollik,

There is limited capability in the exposed API's to do this as of today.  It is expected this is handled in the upstream geometry tools.  The below example shows how you can use the get_bounding_box_of _zonelets method using meshed topo faces as the starting point.  Here we create new labels "my_label" in each part with topo faces that have a bounding box with an xmin > 0. 

Thanks.

```
from ansys.meshing import prime
from ansys.meshing.prime import lucid

prime_client = prime.launch_prime(
    prime_root=r"C:\Program Files\ANSYS Inc\v242\meshing\Prime"
)

model = prime_client.model
mesh_util = lucid.Mesh(model)

pipe_tee = prime.examples.download_pipe_tee_fmd()
mesh_util.read(pipe_tee)

mesh_util.surface_mesh(max_size=10)
message = ""
for part in model.parts:
    for topo_face in part.get_topo_faces():
        meshed_topo_face_zonelet = model.topo_data.get_mesh_zonelets_of_topo_faces(
            [topo_face]
        )
        bbox = prime.SurfaceUtilities(model).get_bounding_box_of_zonelets(
            meshed_topo_face_zonelet
        )
        print(
            f"topo face: {topo_face}\nmeshed zonelet of topo face: {meshed_topo_face_zonelet}\nbbox:\n{bbox}\n",
        )
        if bbox.xmin > 0:
            part.add_labels_on_topo_entities(["my_label"], [topo_face])
            message += (
                f"\nLabel added to topo face '{topo_face}' in part '{part.name}'."
            )
    model.topo_data.delete_mesh_on_topo_faces(
        part.get_topo_faces(),
        prime.DeleteMeshParams(model, delete_mesh_on_connected_topo_edges=True),
    )

print(model, "\n\nTopo faces found with xmin > 0:\n" + message)

```

**Reply by dkowollik** [reply]
Hi @waltersma,
Thank you for your answer. I do not expect all geometry api features to be present. I am just looking for the basic geometry access like connectivities. In your example I first have to create an arbitrary mesh. Is it possible to get access to the edges and nodes of the topo faces? If I could get access to the connectivity of the topo faces, I can loop these entities and identify the named selections. In Spaceclaim I used search algorithms, where I looped the corners and edges of the trimmed nurbs surfaces in order to identify the named selections in a similar manner. 
For my understanding with CAD background, are topo faces like brep faces and face zones like breps?

**Reply by waltersma** [reply]
Hi @dkowollik,

For some topology connectivity queries you can make use of [Topo Data](https://prime.docs.pyansys.com/version/dev/api/_autosummary/ansys.meshing.prime.TopoData.html).  As discussed, for these types of geometry operations I would point you towards pyansys geometry at this stage.  I will add to enhancement requests for these in the issues for future in pyprimemesh.

Topo faces are the geometry faces that are imported from CAD.  These are faceted geometry on import although the nurbs definitions can be maintained for mesh projection.  Face zonelets contain the surface mesh and are typically created from meshing the topo faces.  Face zones are named groups of face zonelets that can be used for applying boundary conditions in solvers.  

Please take a look at the [concepts section](https://prime.docs.pyansys.com/version/dev/user_guide/concepts.html) and this [example](https://prime.docs.pyansys.com/version/stable/examples/gallery_examples/gallery/00_lucid_file_IO.html#sphx-glr-examples-gallery-examples-gallery-00-lucid-file-io-py) for more information about the data structure and how it is used.

Thanks

**Reply by dkowollik** [reply]
Hi @waltersma,

Thank you. I already had a look at the concepts section before. I have named zones without face zonelets (no mesh) in my model. I also had a look at the Topo Data class. There I can get access to the ids of topo faces and topo edges. How can I get access to the corners of the topo edges?

**Reply by waltersma** [reply]
Hi @dkowollik, this is not currently possible. To complete the loop I add the enhancement request link here #880.
